### PR TITLE
SC-24 # Creating serverless project from configuration based routes

### PR DESCRIPTION
**Pull request by mymattcarroll**
### Added
- SC-24: `bm server scope` cmd to display and set scope for project
- SC-24: `bm server info` display of scope
### Changed
- SC-24: `bm server serverless` to allow for configuration based routes as well as directory based routes
### Code Review Notes
- Probably worth reviewing these one commit at a time
- Known incomplete work to be tackled in future:
  - [ ] Write test cases for `lib/serverless.js`

### CODE REVIEW

**Comment by jokeyrhyme** [comment]
For the scope stuff, did you take a look at https://github.com/blinkmobile/blinkmrc.js ? This is what we're using for the BMP CLI and others...

**Comment by mymattcarroll** [comment]
Yeah, thats what it is using at the moment. I tried to follow similar patterns.

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@blinkmobile/server-cli",
-  "description": "develop, test and deploy server-side HTTPS endpoints with BlinkMobile",
+  "description": "CLI to develop, test and deploy server-side HTTPS endpoints with BlinkMobile",
```
Comment (jokeyrhyme) [inline package.json:3]: Try to keep package.json:description the same as that description line near the top of README.md

File: lib/scope.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+function write (
+  cwd /* : string */,
+  meta /* : any */
+) /* : Promise<any> */ {
+  meta = meta || {}
+  if (!meta.project) {
+    return Promise.reject(new Error('meta.project was not defined.'))
+  }
+
+  return projectMeta.write(cwd, (config) => objectMerge(config, {
```
Comment (jokeyrhyme) [inline lib/scope.js:53]: Can you get away with just using `Object.assign()` here instead? That's available in Node.js 4.x and up.

File: package.json (hunk 1 above)
Comment (mymattcarroll) [reply package.json:3]: Will do.

File: lib/scope.js (hunk 2 above)
Comment (mymattcarroll) [reply lib/scope.js:53]: Changing to `Object.assign()` breaks my tests, it does not seem to merge nested properties.

Comment (jokeyrhyme) [reply lib/scope.js:53]: Yeah, I suspected that you needed nested merging. I was just hoping you didn't. :)

**Comment by jokeyrhyme** [comment]
@mymattcarroll the proof-of-concept copied the project to a temporary directory, transformed it, and deployed it.

With the new `copyProject()` and `getProjectPath()` helper functions, what does this process and temporary directory structure look like now?

**Comment by mymattcarroll** [comment]
It still copies the project to a directory specified with the `--out path/to/temp` flag, now it just creates a new directory inside `path/to/temp/project`. This was just to ensure we wont override any files in there project directory. E.g. If they have a file called `handler.js` in the root of there project (however unlikely), it would be overridden with the `handler.js` file we create.

**Comment by jokeyrhyme** [comment]
Ah, okay. Makes sense.

LGTM 👍

**Pull request merged**